### PR TITLE
fix(router) wildcard host matching not considered with plain host match

### PR DESCRIPTION
**Pull request by bungle**
### Summary

When two routes are specified with wildcard host and plain host, the wildcard host may be skipped in some cases. See the test suite change that hilights the problem.

### Issues resolved

Fixes issue reported here:
https://github.com/Kong/kong/issues/3094#issuecomment-363659917

### CODE REVIEW

**Comment by thibaultcha** [comment]
It seems like a couple of wildcard routing unit tests are currently failing in the current CI run?

**Comment by bungle** [comment]
@thibaultcha thanks, it should be fixed now.

File: kong/router.lua [hunk 1]
```diff
@@ -986,23 +988,24 @@ function _M.new(routes)
 
     -- host match
 
-    if plain_indexes.hosts[ctx.req_host] then
-      req_category = bor(req_category, MATCH_RULES.HOST)
```
Comment (thibaultcha) [inline kong/router.lua]: This change can be highly detrimental in many scenarios.

File: kong/router.lua [hunk 2]
```diff
@@ -512,7 +512,8 @@ end
 do
   local matchers = {
     [MATCH_RULES.HOST] = function(route_t, ctx)
-      local host = route_t.hosts[ctx.hits.host or ctx.req_host]
+      local host = route_t.hosts[ctx.req_host] or
+                   route_t.hosts[ctx.hits.host]
```
Comment (thibaultcha) [inline kong/router.lua]: Here, we are ignoring the matched hint, which is thus rendered useless. In which case we could simply remove it. This change can be detrimental in some scenarios.

File: spec/02-integration/05-proxy/01-router_spec.lua [hunk 3]
```diff
@@ -882,6 +882,63 @@ for _, strategy in helpers.each_strategy() do
...
+            ["kong-debug"] = 1,
+          }
+        })
+
+        assert.res_status(200, res)
+
+        assert.equal(routes[2].id,           res.headers["kong-route-id"])
+        assert.equal(routes[2].service.id,   res.headers["kong-service-id"])
+        assert.equal(routes[2].service.name, res.headers["kong-service-name"])
+      end)
```
Comment (thibaultcha) [inline spec/02-integration/05-proxy/01-router_spec.lua]: This test is a duplicate of [this unit test](https://github.com/Kong/kong/blob/master/spec/01-unit/08-router_spec.lua#L519).

File: kong/router.lua (hunk 2 above)
Comment (bungle) [reply kong/router.lua]: What do you mean? If you remove `route_t.hosts[ctx.hits.host]` the tests will not pass, and if you remove `route_t.hosts[ctx.req_host]` the tests won't pass either. So it is not ignored.

**Comment by thibaultcha** [comment]
Replaced by https://github.com/Kong/kong/pull/4775

**Pull request closed without merging**